src/realm/util/sensitive_buffer.hpp

```diff
@@ -0,0 +1,182 @@
+///////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2024 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <cstddef>
+#include <cstring>
+#include <type_traits>
+#include <utility>
+
+#include <realm/util/assert.hpp>
+
+namespace realm::util {
+struct SensitiveBufferBase {
+public:
+    bool engaged() const
+    {
+        return m_buffer != nullptr;
+    }
+
+    size_t size() const
+    {
+        return m_size;
+    }
+
+    virtual ~SensitiveBufferBase();
+
+    bool operator==(const SensitiveBufferBase& rhs) const
+    {
+        if (this == &rhs)
+            return true;
+
+        if (m_size != rhs.m_size)
+            return false;
+
+        if (m_buffer == rhs.m_buffer)
+            return true;
+
+        if (m_buffer == nullptr || rhs.m_buffer == nullptr)
+            return false;
+
+        return std::memcmp(m_buffer, rhs.m_buffer, m_size) == 0;
+    }
+
+    bool operator!=(const SensitiveBufferBase& rhs) const
+    {
+        return !operator==(rhs);
+    }
+
+    static void secure_erase(void*, size_t);
+
+protected:
+    SensitiveBufferBase(size_t buffer_size);
+    SensitiveBufferBase(const SensitiveBufferBase& other);
+    SensitiveBufferBase(SensitiveBufferBase&& other) noexcept;
+
+    template <typename Func>
+    void with_unprotected_buffer(Func f) const
+    {
+        if (!m_buffer) {
+            return;
+        }
+
+        unprotect();
+        f(m_buffer);
+        protect();
```

Comment:
We need to access the encryption key from different threads, so you'd probably have to guard this sequence with a mutex. Further, from what I can tell, a single call to `munlock()` will undo several calls to `mlock()` so we'd have to be very careful to not nest call to get the key anyways; this all seems unsafe.

Reply:
protect and unprotect are noops on non-Windows - mlock and munlock are called when allocating and freeing the buffer only, not during normal access. Besides, this method is protected and only ever used by the templated subclass, so I don’t think there is a way to nest these calls.
But yes, if the buffer is accessed concurrently on Windows this will potentially mess up the flow of CryptUnprotectMemory and CryptProtectMemory calls. I wasn’t sure if the buffer needs to be thread-safe so I didn’t add any synchronization here. I’m surprised our tests pass on Windows (even when ran with encrypted enforced), though perhaps we don’t have a test that would reliably trigger concurrent access?